src/khan.ts

```diff
@@ -0,0 +1,133 @@
+import net from 'net'
+import { Cell, jam, dwim, Noun, enjs } from '@urbit/nockjs'
+import { newtDecode, newtEncode } from './bytes'
+
+// returns a js Noun Object with the structure of a %fyrd
+function Fyrd(desk: string, thread: string, mark_out: string, mark_in: string, atom: string) {
+  return dwim([
+    0,
+    'fyrd',
+    desk,
+    thread,
+    mark_out,
+    mark_in,
+    atom
+  ]);
+}
+
+// returns a js Fyrd, whose trailing atom is
+// some hoon code to be run as a thread
+function Thread (hoon: string) {
+  return Fyrd('base', 'khan-eval', 'noun', 'ted-eval', hoon);
+}
+
+interface ThreadResponse {
+  mark: string,
+  node: {
+    [key: string]: {
+      mark: string;
+      node: Noun;
+    }
+  }
+}
+
+function unwrapUnit(noun: Noun): Noun | null {
+  if (noun instanceof Cell && noun.head.equals(dwim(0))) {
+    return noun.tail;
+  }
+
+  return null;
+}
+
+// decodes response and unpacks it to a noun,
+// if it has the expected structure of a fyrd response
+function parseThreadResponse (data: Buffer): Noun | null {
+  const noun = newtDecode(data);
+  const unwrapped = unwrapUnit(noun);
+  const inner = enjs.pairs([
+    { nom: 'mark', get: enjs.cord },
+    { nom: 'node', get: (n) => n as any }
+  ]);
+  const result =
+      enjs.pairs([
+        { nom: 'mark', get: enjs.cord},
+        {
+          nom: 'node',
+          get: enjs.frond([
+            { tag: '\x01', get: inner },
```

Comment:
yeah, I think there is a missing layer this could be moved down to, ideally devs reading or maintaining this file can be fully insulated from having to think about bytes/hex etc.